bootstrap/bootstrap.go

```diff
@@ -0,0 +1,317 @@
+	"github.com/snapcore/snapd/progress"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/store"
+)
+
+type headerYaml struct {
+	Bootstrap bootstrapYaml `yaml:"bootstrap"`
+}
+
+type bootstrapYaml struct {
+	Snaps            []string `yaml:"extra-snaps"`
+	Rootdir          string   `yaml:"rootdir"`
+	Channel          string   `yaml:"channel"`
+	ModelAssertionFn string   `yaml:"model-assertion"`
+}
+
+func Bootstrap(bootstrapYaml string) error {
+	yamlData, err := ioutil.ReadFile(bootstrapYaml)
+	if err != nil {
+		return err
+	}
+
+	var y headerYaml
+	err = yaml.Unmarshal(yamlData, &y)
+	if err != nil {
+		return err
+	}
+
+	if y.Bootstrap.Rootdir != "" {
+		dirs.SetRootDir(y.Bootstrap.Rootdir)
+		defer dirs.SetRootDir("/")
+	}
+
+	// sanity check target
+	if osutil.FileExists(dirs.SnapStateFile) {
+		return fmt.Errorf("cannot bootstrap over existing system")
+	}
+
+	rawAssert, err := ioutil.ReadFile(y.Bootstrap.ModelAssertionFn)
+	if err != nil {
+		return err
+	}
+
+	ass, err := asserts.Decode(rawAssert)
+	if err != nil {
+		return err
+	}
+	model := ass.(*asserts.Model)
+
+	// *sigh* we need to adjust the storeID if its set to "canonical"
+	//        because that is the default
+	storeID := model.Store()
+	if storeID == "canonical" {
+		storeID = ""
+	}
+
+	// put snaps in place
+	if err := os.MkdirAll(dirs.SnapBlobDir, 0755); err != nil {
+		return err
+	}
+	opts := &downloadOptions{
+		TargetDir:    dirs.SnapBlobDir,
+		Channel:      y.Bootstrap.Channel,
+		StoreID:      storeID,
+		Architecture: model.Architecture(),
+	}
+
+	// FIXME: support sideloading snaps by copying the boostrap.snaps
+	//        first and keeping track of the already downloaded names
+	snaps := []string{}
+	snaps = append(snaps, y.Bootstrap.Snaps...)
+	snaps = append(snaps, model.Gadget())
+	snaps = append(snaps, model.OS())
+	snaps = append(snaps, model.Kernel())
+	snaps = append(snaps, model.RequiredSnaps()...)
+
+	for _, snapName := range snaps {
+		fmt.Printf("Fetching %s\n", snapName)
+		if osutil.FileExists(snapName) {
+			if err := copyLocalSnapFile(snapName); err != nil {
+				return err
+			}
+		} else {
+			if _, err := downloadSnapWithSideInfo(snapName, opts); err != nil {
+				return err
+			}
+		}
+	}
+
+	// now do the bootloader stuff
+	if err := setBootvars(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func setBootvars() error {
+	// set the bootvars for kernel/os snaps so that the system
+	// actually boots and can do the `firstboot` import of the snaps.
+	//
+	// there is also no mounted os/kernel snap, all we have are the
+	// blobs
+
+	bootloader, err := partition.FindBootloader()
+	if err != nil {
+		return fmt.Errorf("can not set kernel/os bootvars: %s", err)
+	}
+
+	snaps, _ := filepath.Glob(filepath.Join(dirs.SnapBlobDir, "*.snap"))
+	if len(snaps) == 0 {
+		return fmt.Errorf("internal error: cannot find os/kernel snap")
+	}
+	for _, fullname := range snaps {
```

Comment:
this seems wasteful, we know what are the relevant snaps from the config and we can capture enough information in some kind of table from the download/copy phase not to need to reopen snaps here